### PR TITLE
Predefined various timestamp formats

### PR DESCRIPTION
**Pull request by ghost**
This extension allows you to change the format of the timestamp when using `fping -D -l 127.0.0.1`.
Example: `fping -D -l --timestamp-format=ctime 127.0.0.1`

Issue: #310

### CODE REVIEW

File: src/fping.c [hunk 1]
```diff
@@ -2992,6 +3036,7 @@ void usage(int is_error)
     fprintf(out, "   -C, --vcount=N     same as -c, report results (not stats) in verbose format\n");
     fprintf(out, "   -d, --rdns         show targets by name (force reverse-DNS lookup)\n");
     fprintf(out, "   -D, --timestamp    print timestamp before each output line\n");
+    fprintf(out, "       --timestamp-format=FORMAT  show timestamp using the given format (-D are required): [ctime|iso|reltime]\n");
```
Comment (schweikert) [inline src/fping.c]: `-D is required` or `-D required` to shorten it.

Comment (schweikert) [reply src/fping.c]: To shorten it a bit more:

show timestamp in the given format (-D required): ctime|iso|reltime\n

So maybe wrapping can be avoided.

**Review by schweikert — APPROVED** [review]

File: src/fping.c (hunk 1 above)
Comment (gsnw) [reply src/fping.c]: Thanks for the feedback and the suggestion.
I have adopted the change and hope the rest of the code is also ok

File: ci/test-05-options-c-e.pl [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/perl -w
 
-use Test::Command tests => 33;
+use Test::Command tests => 42;
```
Comment (auerswal) [inline ci/test-05-options-c-e.pl]: Please consider adding two error test cases for the new option:

1. omitting the format name, and
2. using an unsupported format name.

File: src/fping.c [hunk 3]
```diff
@@ -2939,6 +2955,34 @@ void ev_remove(struct event_queue *queue, struct event *event)
...
+{
+    char time_buffer[100];
+    time_t current_time_s;
+    struct tm *local_time;
+
+    if(timestamp_format >= 1 && timestamp_format <= 3) {
+        current_time_s = current_time_ns / 1000000000;
+        local_time = localtime(&current_time_s);
+        if(timestamp_format == 1) {
+          strftime(time_buffer, sizeof(time_buffer), "%c", local_time);
```
Comment (auerswal) [inline src/fping.c]: Before this patch, `strftime` is not used in the `fping` code. Should `configure` check for availability of `strftime`?

File: src/fping.c [hunk 4]
```diff
@@ -2939,6 +2955,34 @@ void ev_remove(struct event_queue *queue, struct event *event)
...
+void print_timestamp_format(int64_t current_time_ns, int timestamp_format)
+{
+    char time_buffer[100];
+    time_t current_time_s;
+    struct tm *local_time;
+
+    if(timestamp_format >= 1 && timestamp_format <= 3) {
+        current_time_s = current_time_ns / 1000000000;
+        local_time = localtime(&current_time_s);
+        if(timestamp_format == 1) {
```
Comment (auerswal) [inline src/fping.c]: I'd think this part (i.e., the inner `if`) would be more readable using `switch`.

**Review by auerswal — APPROVED** [review]
All in all, this looks OK. Please consider my comments, but you do not need to make any changes on my behalf.

Thanks!

**Comment by auerswal** [comment]
Please also consider adding a description to the man page, i.e., `fping.pod`. Thanks!

**Comment by auerswal** [comment]
I am not so sure any more, especially regarding `reltime`. With `dmesg`, `--time-format=reltime` provides *relative* time stamps for short time differences, and absolute time stamps for longer gaps, e.g.:

```
[Jun 9 21:17] …
[ +21,801486] …
[  +0,000004] …
[  +0,000004] …
[  +0,000002] …
[  +0,000001] …
[  +0,091934] …
[  +0,000003] …
[  +0,000002] …
[  +0,000002] …
[  +0,000001] …
[  +0,000004] …
[  +0,574391] …
[Jun 9 22:54] …
```

Perhaps we should leave out `reltime` for now?

And a nit: both `date --iso-8601` and `dmesg --time-format=iso` print the timezone offset with a colon between hours and minutes (`HH:MM`). The `%z` format specifier does not include the colon, but I did not see a better format specifier in `man strftime`.

(As another example, GNU date has two quite similar "named" time format options where the output differs mainly with the character between date and time, either a space or a `T`: `date --rfc-3339=s` and `date --iso-8601=s` [`--rfc-3339` does not use the `locale` settings, `--iso-8601` does, but `fping` does not use `locale` settings (IMHO `fping` is better in that regard)].)

File: ci/test-05-options-c-e.pl (hunk 2 above)
Comment (gsnw) [reply ci/test-05-options-c-e.pl]: I have add two negativ test `--timestamp-format` and `--timestamp-format="%Y-%m-%d %H:%M:%S"` both ending in function usage() with exit code 1

File: src/fping.c (hunk 4 above)
Comment (gsnw) [reply src/fping.c]: The if statement has been completely converted into a switch statement. Many thanks for this feedback.
Switch statement with the default is also better than the old if else.
This way, the default is always used if the function is passed incorrectly

**Comment by gsnw** [comment]
The formats are not perfect. I have now turned reltime into rfc3339, even with the knowledge that the format may not be one hundred percent perfect.
However, it corresponds to the example from the ticket.

[1] https://medium.easyread.co/understanding-about-rfc-3339-for-datetime-formatting-in-software-engineering-940aa5d5f68a?gi=fc35d513c0fd

File: src/fping.c (hunk 3 above)
Comment (gsnw) [reply src/fping.c]: It's also done

**Review by auerswal — APPROVED** [review]
Thanks for the adjustments, this looks good. :-)

**Comment by auerswal** [comment]
We haven't talked about the process for merging pull requests yet. I would be willing to merge this pull request next weekend, unless there are objections. Is that OK?

**Comment by gsnw** [comment]
The last github merges were probably done with rebase.
Everything has advantages and disadvantages.

All my pull requests are always based on the current state of develop and I always try to do a rebase on all branches in my repository when the develop branch changes here.

So I would prefer rebase and merge

**Comment by auerswal** [comment]
Ah, I see, that is one of the options shown when clicking the arrow next to *Merge pull request*. I think I (slightly) prefer that option too, because automatic merge commits add some noise, but only little information. I'll go ahead and do that, thanks!

(For the record: I do *not* like the *Squash and merge* option, I prefer to keep the individual commits.)

**Pull request merged**